File: src/main/java/cz/encircled/eplayer/common/PostponeTimer.java
package cz.encircled.eplayer.common;

import java.util.Timer;
import java.util.TimerTask;

/**
 * @author Encircled on 28/09/2014.
 */
public class PostponeTimer {

    private final Runnable runnable;

    private Timer timer;

    public PostponeTimer(Runnable runnable) {
        this.runnable = runnable;
    }

    public void postpone(long delay) {
        reset(delay);
    }

    public void cancel() {
        if (timer != null)
            timer.cancel();
    }

    private void reset(long delay) {
        cancel();
        timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                runnable.run();
            }
        }, delay);
    }

}

File: src/main/java/cz/encircled/eplayer/common/PostponeTimer.kt
package cz.encircled.eplayer.common

import java.lang.Runnable
import java.util.*

/**
 * @author Encircled on 28/09/2014.
 */
class PostponeTimer(delay: Long? = null, private val runnable: Runnable) {

    private var timer: Timer? = null

    init {
        delay?.let { postpone(it) }
    }

    fun postpone(delay: Long) {
        reset(delay)
    }

    fun cancel() {
        if (timer != null) timer!!.cancel()
    }

    private fun reset(delay: Long) {
        cancel()
        timer = Timer()
        timer!!.schedule(object : TimerTask() {
            override fun run() {
                runnable.run()
            }
        }, delay)
    }
}